common/thorhelper/thorread.cpp

```diff
@@ -0,0 +1,825 @@
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "jliball.hpp"
+
+#include "thorfile.hpp"
+
+#include "eclhelper.hpp"
+#include "eclrtl.hpp"
+#include "eclrtl_imp.hpp"
+#include "rtlfield.hpp"
+#include "rtlds_imp.hpp"
+#include "rtldynfield.hpp"
+
+#include "rmtclient.hpp"
+#include "rmtfile.hpp"
+
+#include "thorread.hpp"
+#include "rtlcommon.hpp"
+#include "thorcommon.hpp"
+
+//---------------------------------------------------------------------------------------------------------------------
+
+static Owned<IPropertyTree> emptyTree = createPTree();
+
+//---------------------------------------------------------------------------------------------------------------------
+
+constexpr size32_t defaultReadBufferSize = 0x10000;
+
+//Cannot use InterfaceOf<> because multiple base interfaces are derived from IInterface
+class DiskRowReader : implements IAllocRowStream, implements IRawRowStream, implements IDiskRowReader, implements IThorDiskCallback, public CInterface
+{
+public:
+    DiskRowReader(unsigned _expectedCrc, IOutputMetaData & _expected, unsigned _projectedCrc, IOutputMetaData & _projected, unsigned _actualCrc, IOutputMetaData & _actual, IPropertyTree * options);
+    IMPLEMENT_IINTERFACE
+
+    virtual IRawRowStream * queryRawRowStream() override;
+    virtual IAllocRowStream * queryAllocatedRowStream(IEngineRowAllocator * _outputAllocator) override;
+
+    virtual bool getCursor(MemoryBuffer & cursor) override;
+    virtual void setCursor(MemoryBuffer & cursor) override;
+    virtual void stop() override;
+
+    virtual void clearInput() override;
+    virtual bool matches(const char * format, bool streamRemote, unsigned _expectedCrc, IOutputMetaData & _expected, unsigned _projectedCrc, IOutputMetaData & _projected, unsigned _actualCrc, IOutputMetaData & _actual, IPropertyTree * options) override;
+
+// IThorDiskCallback
+    virtual offset_t getFilePosition(const void * row) override;
+    virtual offset_t getLocalFilePosition(const void * row) override;
+    virtual const char * queryLogicalFilename(const void * row) override;
+    virtual const byte * lookupBlob(unsigned __int64 id) override { UNIMPLEMENTED; }
+
+
+protected:
+    offset_t getLocalOffset();
+
+protected:
+    Owned<ISerialStream> input;
+    Owned<IFileIO> inputfileio;
+    CThorContiguousRowBuffer inputBuffer;
+    Owned<IEngineRowAllocator> outputAllocator;
+    IOutputMetaData * expectedDiskMeta = nullptr;
+    IOutputMetaData * projectedDiskMeta = nullptr;
+    IOutputMetaData * actualDiskMeta = nullptr;
+    unsigned expectedCrc;
+    unsigned projectedCrc;
+    unsigned actualCrc;
```

Comment:
coding style question: should be initialize all members as a matter of good practice, even if they are also initialize in ctor.
I think for members that have ctor initializers - compiler guarantees won't do twice.